pkg/vminetworkscontroller/vmi_controller.go

```diff
@@ -202,29 +198,42 @@ func (r *VirtualMachineInstanceReconciler) vmiNetworksClaimingIPAM(
 	return vmiNets, nil
 }
 
-func (r *VirtualMachineInstanceReconciler) Cleanup(vmiKey apitypes.NamespacedName) error {
-	ipamClaims := &ipamclaimsapi.IPAMClaimList{}
-	listOpts := []client.ListOption{
-		client.InNamespace(vmiKey.Namespace),
-		ownedByVMLabel(vmiKey.Name),
-	}
-	if err := r.Client.List(context.Background(), ipamClaims, listOpts...); err != nil {
-		return fmt.Errorf("could not get list of IPAMClaims owned by VM %q: %w", vmiKey.String(), err)
+func shouldCleanFinalizers(vmi *virtv1.VirtualMachineInstance, vm *virtv1.VirtualMachine) bool {
+	return (vm != nil && vmi == nil && vm.DeletionTimestamp != nil) ||
+		(vm == nil && (vmi == nil || (vmi.DeletionTimestamp != nil && len(vmi.Status.ActivePods) == 0)))
+}
+
+func ownerReferenceFor(vmi *virtv1.VirtualMachineInstance, vm *virtv1.VirtualMachine) metav1.OwnerReference {
+	var obj client.Object
+	if vm != nil {
+		obj = vm
+	} else {
+		obj = vmi
 	}
 
-	for _, claim := range ipamClaims.Items {
-		removedFinalizer := controllerutil.RemoveFinalizer(&claim, kubevirtVMFinalizer)
-		if removedFinalizer {
-			if err := r.Client.Update(context.Background(), &claim, &client.UpdateOptions{}); err != nil {
-				return client.IgnoreNotFound(err)
-			}
-		}
+	aPIVersion := obj.GetObjectKind().GroupVersionKind().Group + "/" + obj.GetObjectKind().GroupVersionKind().Version
+	return metav1.OwnerReference{
+		APIVersion:         aPIVersion,
+		Kind:               obj.GetObjectKind().GroupVersionKind().Kind,
+		Name:               obj.GetName(),
+		UID:                obj.GetUID(),
+		Controller:         ptr.To(true),
+		BlockOwnerDeletion: ptr.To(true),
```

Comment:
There is a tool at controller-runtime to mark owners as controllers, let's use it.
```golang
sigs.k8s.io/pkg/controller/controllerutil.SetControllerReference
```

Reply:
I prefer please to do it on a follow-up, i did in this PR the minimal changes needed for finalizers
it is better to have PRs focused and not massive

Reply:
Ok, let's do a follow up, it will be something like

```golang
controllerutil.SetControllerReference(vm, vmi, scheme.Scheme)
```

Reply:
yep, it will need to know both cases of VMI owner and VM owner